File: src/go/challenges/problems/final_value_of_variable_after_performing_operations_test.go
// https://leetcode.com/problems/final-value-of-variable-after-performing-operations

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type FinalValueOfVariableAfterPerformingOperations struct{}

func TestFinalValueOfVariableAfterPerformingOperations(t *testing.T) {
	gen := core.TestSuite[FinalValueOfVariableAfterPerformingOperations]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([]string{"--X", "X++", "X++"}).Result(1)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]string{"++X", "++X", "X++"}).Result(3)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]string{"X++", "++X", "--X", "X--"}).Result(0)
	}).Run(t)
}

func (FinalValueOfVariableAfterPerformingOperations) Solution(operations []string) int {
	x := 0
	for _, op := range operations {
		if op == "++X" || op == "X++" {
			x++
		} else {
			x--
		}
	}

	return x
}

File: src/python/challenges/problems/final_value_of_variable_after_performing_operations_test.py

# https://leetcode.com/problems/final-value-of-variable-after-performing-operations

from typing import List
from core.problem_base import *

class FinalValueOfVariableAfterPerformingOperations(ProblemBase):
    def Solution(self, operations: List[str]) -> int:
        operations_map = {
            "++X": 1,
            "X++": 1,
            "--X": -1,
            "X--": -1
        }

        X = 0
        for _, op in enumerate(operations):
            if op in operations_map:
                X += operations_map[op]

        return X

if __name__ == '__main__':
    TestGen(FinalValueOfVariableAfterPerformingOperations) \
        .Add(lambda tc: tc.Param(["--X","X++","X++"]).Result(1)) \
        .Add(lambda tc: tc.Param(["++X","++X","X++"]).Result(3)) \
        .Add(lambda tc: tc.Param(["X++","++X","--X","X--"]).Result(0)) \
        .Run()
